src/java/org/apache/cassandra/service/DataResolver.java

```diff
@@ -150,46 +150,46 @@ private PartitionIterator resolveWithReplicaFilteringProtection()
         // at the beginning of this method), so grab the response count once and use that through the method.
         int count = responses.size();
         // We need separate contexts, as each context has his own counter
         ResolveContext firstPhaseContext = new ResolveContext(count);
         ResolveContext secondPhaseContext = new ResolveContext(count);
-        ReplicaFilteringProtection rfp = new ReplicaFilteringProtection(keyspace, command, consistency, firstPhaseContext.sources);
+
+        ReplicaFilteringProtection rfp = new ReplicaFilteringProtection(keyspace,
+                                                                        command,
+                                                                        consistency,
+                                                                        firstPhaseContext.sources,
+                                                                        DatabaseDescriptor.getCachedReplicaRowsWarnThreshold(),
+                                                                        DatabaseDescriptor.getCachedReplicaRowsFailThreshold());
+
         PartitionIterator firstPhasePartitions = resolveInternal(firstPhaseContext,
                                                                  rfp.mergeController(),
                                                                  i -> shortReadProtectedResponse(i, firstPhaseContext),
                                                                  UnaryOperator.identity());
-
-        // Consume the first phase partitions to populate the replica filtering protection with both those materialized
-        // partitions and the primary keys to be fetched.
-        PartitionIterators.consume(firstPhasePartitions);
-        firstPhasePartitions.close();
-
-        // After reading the entire query results the protection helper should have cached all the partitions so we can
-        // clear the responses accumulator for the sake of memory usage, given that the second phase might take long if
-        // it needs to query replicas.
-        responses.clearUnsafe();
-
-        return resolveWithReadRepair(secondPhaseContext,
-                                     rfp::queryProtectedPartitions,
-                                     results -> command.rowFilter().filter(results, command.metadata(), command.nowInSec()));
+        
+        PartitionIterator completedPartitions = resolveWithReadRepair(secondPhaseContext,
+                                                                      i -> rfp.queryProtectedPartitions(firstPhasePartitions, i),
+                                                                      results -> command.rowFilter().filter(results, command.metadata(), command.nowInSec()));
+        
+        // Ensure that the RFP instance has a chance to record metrics when the iterator closes.
+        return PartitionIterators.doOnClose(completedPartitions, rfp::close);
```

Comment:
I like the new `doOnClose` method. However, we are not closing the first phase iterator, and I think we should. We could either close it in `ReplicaFilteringProtection#close` or right here with a lambda, but I think it would be easier/cleaner if we move the `ReplicaFilteringProtection#close` method to the `UnfilteredPartitionIterators.MergeListener` returned by `ReplicaFilteringProtection#mergeController()`, and here we simply close the 1st phase iterator:
```suggestion
        return PartitionIterators.doOnClose(completedPartitions, firstPhasePartitions::close);
```
Closing the 1st phase iterator here feels natural IMO, and allows us to abstract the `DataResolver` from the RFP metrics stuff. 

Alternatively, if you don't agree and want to keep the `ReplicaFilteringProtection#close()` method, we could make `ReplicaFilteringProtection` implement `Closeable`, although I'd prefer using the `close()` method of the listener. WDYT?